Refactor plots2d

cwehmeyer opened the pull request:
This PR refactors the plots2d module:

- auxiliary functions allow to make custom plots easily
- new plotting functions (plot_density, plot_free_energy_new, plot_contour) with more control (e.g. color bars)
- old plotting functions retain signatures and behaviour but use new implementation

cwehmeyer committed:
adding tests for new plotting functions

cwehmeyer committed:
implement new plotting functions

cwehmeyer committed:
added .pytest_cache to .gitignore

cwehmeyer committed:
avoiding name clash with old function

cwehmeyer committed:
base old functions on new code

cwehmeyer committed:
fix py2 compatibility and broken import in tests

cwehmeyer committed:
typos

cwehmeyer committed:
added missing import

cwehmeyer committed:
fixing function parameters

cwehmeyer committed:
amended CHANGELOG

marscher commented on doc/source/CHANGELOG.rst:
```diff
@@ -16,6 +16,8 @@ Changelog
 - msm: fixed serialization of BayesianHMSM, if initialized with a ML-HMSM. #1283
 - base: require at least tqdm >= 4.23, because of an API change. #1292, #1293
 - plots: fixed minor bug in plot_network (state_labels=None would not work). #1306
+- plots: refactored plots2d
```
this message is kind of uninformative to the user. I would state, that the gca behaviour has changed etc.

marscher commented on pyemma/plots/plots2d.py:
```diff
@@ -121,23 +114,27 @@ def plot_free_energy(xall, yall, weights=None, ax=None, nbins=100, ncountours=10
         number of histogram bins used in each dimension
     ncountours : int, default=100
         number of contours used
     offset : float, default=0.1
```
default is -1 now.

marscher commented on pyemma/plots/plots2d.py:
```diff
@@ -156,33 +153,403 @@ def plot_free_energy(xall, yall, weights=None, ax=None, nbins=100, ncountours=10
 
     """
-    import matplotlib.pylab as _plt
+    #import matplotlib.pylab as _plt
```
delete me

marscher commented on pyemma/plots/plots2d.py:
```diff
@@ -156,33 +153,403 @@ def plot_free_energy(xall, yall, weights=None, ax=None, nbins=100, ncountours=10
+        logscale=logscale)
+    
+
+def plot_free_energy_new(
```
what is actually the difference to the old method? Can we decide to have only one or are they fundamentally different?
A short line in the docstring what this method does different would be nice.

cwehmeyer replied:
The new plotting functions always return `fig, ax, cbar` while the old functions return either `fig, ax` or just `ax`. Further, `plot_free_energy` has a parameter typo (`countour`) which has been fixed in the new function.

I would love to immediately replace everything with the new functions but this would slightly break the API.

marscher replied:
We could have a legacy switch parameter to steer the old or new behaviour, but this would convolute the code a bit. I think we should deprecate the old function and promote the new one, shouldn't we?

cwehmeyer replied:
Yes, indeed. I don't think functions would be too convoluted, though. Using a legacy switch is a good idea!

cwehmeyer replied:
Do we need to conserve the sequence of named parameters in legacy mode?

marscher reviewed (approved):
Thank you very much for cleaning this up! Especially not using gca in cases where it is inappropriate is a big win.

cwehmeyer committed:
added change requests from review

cwehmeyer committed:
deprecate old functions; add legacy mode for plot_free_energy; add plot_scatter_contour function

marscher commented:
> 
> Do we need to conserve the sequence of named parameters in legacy mode?
> 

strictly speaking yes, because if somebody put in the parameters in the
previous order, you would break the intended usage of the parameter.

cwehmeyer committed:
preserve parameter sequence in plot_free_energy

cwehmeyer committed:
fixed function name in tests

cwehmeyer committed:
importing+documenting new plotting functions

cwehmeyer committed:
bugfix in plot_map for vmin=None and/or vmax=None in case of infinity values

cwehmeyer committed:
remove implicit vmin/vmax setting in plot_map

thempel committed:
[plots1d] eliminate plt.gca(), fix test class name typo

cwehmeyer committed:
amending parameter lists and function calls

cwehmeyer committed:
Merge branch 'refactor_plots2d' of github.com:markovmodel/PyEMMA into refactor_plots2d

cwehmeyer committed:
removed obsolete code

cwehmeyer committed:
added missing parameter to signature

cwehmeyer committed:
added missing parameter to signature

cwehmeyer committed:
variable name bugfix

cwehmeyer committed:
minor fixes

cwehmeyer commented:
@thempel please have a look at `test_feature_histograms_nowarning` in `TestPlots1d`.

thempel committed:
[plots1d] deleted redundant test, added tests for function arguments

Test for plotting function to not raise a warning failed on
travis / CI probably because of "RuntimeWarning: More than 20
figures have been opened [...]". Test removed because
non-informative (did not display warning), same warning would be
raised in test_feature_histograms.

thempel commented:
Got it. This test was too unspecific and pretty much redundant (checked if no warning is raised for this particular function). Also, it did not show the raised warning. I removed the test and put some more meaningful ones there instead. 
Anyway, probably the failure was caused by this warning. Don't know if it's important, but just for the record.

RuntimeWarning: More than 20 figures have been opened. Figures created through the pyplot interface (`matplotlib.pyplot.figure`) are retained until explicitly closed and may consume too much memory. (To control this warning, see the rcParam `figure.max_open_warning`).     max_open_warning, RuntimeWarning)

cwehmeyer committed:
cleanup

cwehmeyer committed:
removed plot_scatter_contour

cwehmeyer committed:
use default cmap=None

cwehmeyer committed:
replace plot_map's logscale parameter with norm and restrict logscale to plot_density

cwehmeyer committed:
removed obsolete imports, added tests, and plot_contours accepts levels='legacy'

cwehmeyer committed:
minor fixes

thempel commented:
For `plot_density()`, I get different behavior if using logscale or not. Logscale produces a discrete colormap which is otherwise continuous. Was that intended?

thempel committed:
[plots2d] consistent behavior for logscale=True and False

Adjust levels to be equally fine in both cases

thempel commented:
@cwehmeyer we discussed that it is potentially useful to have access to colormap or mappable. Since I found it a bit inconsistent to return either the colormap or the mappable, depending on the kwargs, they are now written into a dictionary and returned together. Otherwise, larger customized plots might break down if somebody chooses to alter some of the plot options. I'm not completely happy with it though; if you have a better idea feel free to implement it.

cwehmeyer committed:
minor changes

cwehmeyer committed:
added plot_state_map function

cwehmeyer committed:
import fix

cwehmeyer committed:
minor change

cwehmeyer committed:
transpose z in get_histogram instead of plot_density and plot_free_energy

cwehmeyer committed:
use histogram-based mask in plot_state_map

cwehmeyer committed:
fix x/y-orientation

thempel committed:
[plots2d] removed ncontours for plot_state_map function

thempel committed:
[plots1d] reversed hist order, new default ylabels, return fig, ax

cwehmeyer committed:
allow kwargs in plot_map and its wrappers to pass optional parameters to contourf

cwehmeyer committed:
close figures in plots2d's tests to prevent too_many_figures warnings

thempel committed:
[plots1d] close figures in tests

thempel committed:
Merge remote-tracking branch 'origin/refactor_plots2d' into refactor_plots2d

cwehmeyer committed:
allow masking of unsampled areas in plot_contour

cwehmeyer committed:
use method='nearest' in plot_contour

cwehmeyer committed:
fixing kwargs filter

thempel commented:
Thx @cwehmeyer. Can you add a line about the kwargs (i.e. what function they are passed to) to the docstrings?

cwehmeyer committed:
amending docstrings

cwehmeyer commented:
@thempel: done.

thempel committed:
[test_plots2d] remove warnings import
